WalletWasabi/WabiSabi/Client/DependencyGraphTaskScheduler.cs

```diff
@@ -158,29 +157,37 @@ public async Task StartOutputRegistrationsAsync(IEnumerable<TxOut> txOuts, BobCl
 				vsizeCredsToPresentTasks,
 				Array.Empty<TaskCompletionSource<Credential>>(),
 				Array.Empty<TaskCompletionSource<Credential>>());
+			return smartRequestNode;
+		});
 
-			var task = smartRequestNode
-				.StartOutputRegistrationAsync(bobClient, txOut.ScriptPubKey, cancellationToken)
-				.ContinueWith((t) =>
+		var remainingTime = outputRegistrationEndTime - DateTimeOffset.UtcNow;
+		if (remainingTime < TimeSpan.FromSeconds(5))
+		{
+			throw new InvalidOperationException("No time to register the outputs, aborting.");
+		}
+
+		var delays = remainingTime.SamplePoissonDelays(txOuts.Count());
+
+		var tasks = txOuts.Zip(nodes, delays,
+			async (txOut, smartRequestNode, delay) =>
+			{
+				try
 				{
-					if (t.IsFaulted && t.Exception is { } exception)
+					await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+					await smartRequestNode.StartOutputRegistrationAsync(bobClient, txOut.ScriptPubKey, cancellationToken).ConfigureAwait(false);
+				}
+				catch (WabiSabiProtocolException ex) when (ex.ErrorCode == WabiSabiProtocolErrorCode.AlreadyRegisteredScript)
+				{
+					if (keyChain is KeyChain { KeyManager: var keyManager }
+						&& keyManager.TryGetKeyForScriptPubKey(txOut.ScriptPubKey, out var hdPubKey))
 					{
-						if (exception.InnerExceptions.Select(e => e.InnerException).Any(x => x is WabiSabiProtocolException wabisabiexc
-							&& wabisabiexc.ErrorCode == WabiSabiProtocolErrorCode.AlreadyRegisteredScript))
-						{
-							if (keyChain is KeyChain { KeyManager: var keyManager }
-								&& keyManager.TryGetKeyForScriptPubKey(txOut.ScriptPubKey, out var hdPubKey))
-							{
-								hdPubKey.SetKeyState(KeyState.Used);
-							}
-						}
-
-						throw exception;
+						hdPubKey.SetKeyState(KeyState.Used);
```

Comment:
It seems to me that `IKeyChain` needs a `MarkAsUsed(Script)` method. This is a hack and if tomorrow someone implements a HardwareWalletKeyChain then we will have problems.  

Anyway, it seems to me that based on Adam recent patch, this job should be done by the WalletManager because it needs to be sore to mark the script as `Used` in **All** the existing wallet to reduce the chances of been reused.

Reply:
Concept ACK, but I think it is not the scope of this PR.